File: go_backend/services/user_service_test.go
package services

import (
	"context"
	"testing"

	"syncflow/entities"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
)

// MockUserDBOps is a mock of UserDBOps
type MockUserDBOps struct {
	mock.Mock
}

func (m *MockUserDBOps) Create(ctx context.Context, user *entities.User) error {
	args := m.Called(ctx, user)
	return args.Error(0)
}

func (m *MockUserDBOps) GetByID(ctx context.Context, id uint) (*entities.User, error) {
	args := m.Called(ctx, id)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*entities.User), args.Error(1)
}

func (m *MockUserDBOps) GetByEmail(ctx context.Context, email string) (*entities.User, error) {
	args := m.Called(ctx, email)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*entities.User), args.Error(1)
}

func TestCreateUser(t *testing.T) {
	mockDBOps := new(MockUserDBOps)
	service := NewUserService(mockDBOps)
	ctx := context.Background()

	t.Run("Success", func(t *testing.T) {
		req := CreateUserRequest{
			Name:  "John Doe",
			Email: "john@example.com",
		}

		mockDBOps.On("GetByEmail", ctx, req.Email).Return(nil, nil).Once()
		mockDBOps.On("Create", ctx, mock.AnythingOfType("*entities.User")).Return(nil).Once()

		user, err := service.CreateUser(ctx, req)

		assert.NoError(t, err)
		assert.NotNil(t, user)
		assert.Equal(t, req.Name, user.Name)
		assert.Equal(t, req.Email, user.Email)
		mockDBOps.AssertExpectations(t)
	})

	t.Run("UserAlreadyExists", func(t *testing.T) {
		req := CreateUserRequest{
			Name:  "John Doe",
			Email: "john@example.com",
		}
		existingUser := &entities.User{Email: req.Email}

		mockDBOps.On("GetByEmail", ctx, req.Email).Return(existingUser, nil).Once()

		user, err := service.CreateUser(ctx, req)

		assert.Error(t, err)
		assert.Nil(t, user)
		assert.Contains(t, err.Error(), "already exists")
		mockDBOps.AssertExpectations(t)
	})
}

File: node_backend/src/services/user_service.test.ts
// services/user_service.test.ts — Equivalent to go_backend/services/user_service_test.go
import { UserServiceImpl, CreateUserRequest } from './user_service';
import { UserDBOps } from '../dbops/user_dbops';
import { User } from '../entities/user';

/**
 * MockUserDBOps — Equivalent to Go's MockUserDBOps using testify/mock.
 */
class MockUserDBOps implements UserDBOps {
    createFn = jest.fn<Promise<User>, [{ name: string; email: string }]>();
    getByIDFn = jest.fn<Promise<User | null>, [number]>();
    getByEmailFn = jest.fn<Promise<User | null>, [string]>();

    async create(data: { name: string; email: string }): Promise<User> {
        return this.createFn(data);
    }

    async getByID(id: number): Promise<User | null> {
        return this.getByIDFn(id);
    }

    async getByEmail(email: string): Promise<User | null> {
        return this.getByEmailFn(email);
    }
}

describe('UserService', () => {
    let mockDBOps: MockUserDBOps;
    let service: UserServiceImpl;

    beforeEach(() => {
        mockDBOps = new MockUserDBOps();
        service = new UserServiceImpl(mockDBOps);
    });

    describe('createUser', () => {
        it('should create a user successfully', async () => {
            const req: CreateUserRequest = {
                name: 'John Doe',
                email: 'john@example.com',
            };

            const createdUser: User = {
                id: 1,
                name: req.name,
                email: req.email,
                createdAt: new Date(),
                updatedAt: new Date(),
                deletedAt: null,
            };

            mockDBOps.getByEmailFn.mockResolvedValue(null);
            mockDBOps.createFn.mockResolvedValue(createdUser);

            const user = await service.createUser(req);

            expect(user).toBeDefined();
            expect(user.name).toBe(req.name);
            expect(user.email).toBe(req.email);
            expect(mockDBOps.getByEmailFn).toHaveBeenCalledWith(req.email);
            expect(mockDBOps.createFn).toHaveBeenCalledWith({
                name: req.name,
                email: req.email,
            });
        });

        it('should throw error when user already exists', async () => {
            const req: CreateUserRequest = {
                name: 'John Doe',
                email: 'john@example.com',
            };

            const existingUser: User = {
                id: 1,
                name: 'Existing',
                email: req.email,
                createdAt: new Date(),
                updatedAt: new Date(),
                deletedAt: null,
            };

            mockDBOps.getByEmailFn.mockResolvedValue(existingUser);

            await expect(service.createUser(req)).rejects.toThrow('already exists');
            expect(mockDBOps.getByEmailFn).toHaveBeenCalledWith(req.email);
            expect(mockDBOps.createFn).not.toHaveBeenCalled();
        });
    });
});
